### ISSUE TITLE
JupyterLab Advanced (JSON) settings editor sometimes does not launch

### ISSUE THREAD

**Issue by jasongrout** [jupyterlab/jupyterlab#12485]
## Description

The JupyterLab JSON advanced settings editor sometimes does not launch.

## Reproduce

1. Install JupyterLab 3.3.4 (macos 12, m1 processor)
2. Open command palette and execute the "Advanced JSON settings editor" command

Sometimes the JSON editor does not appear. For me, in launching jupyterlab 5 times, the command does not work maybe in one of the launches. If the command doesn't work the first time, it seems to not work until jupyterlab is relaunched. Refreshing the page doesn't seem to fix it.

There may be a race condition of some kind? This is puzzling.

This came up in the PyCon US 2022 JupyterLab tutorial.

## Expected behavior

Advanced JSON setting editor always pops up.

## Context

- Operating System and version: MacOS 12
- Browser and version: Firefox 99
- JupyterLab version: 3.3.4

I don't see any javascript errors in the browser console.

CC @pya

**Comment by echarles** [comment]
I have faced one time that issue while working today on https://github.com/jupyterlab/jupyterlab/pull/12466 I have opened/refreshed numerous times (>50 ? <100) the editor, and faced that issue one time. The linked PR will not help to solve that.

**Comment by echarles** [comment]
See comment on https://github.com/jupyterlab/jupyterlab/pull/12470#issuecomment-1112942395 cc/ @fcollonval

**Comment by echarles** [comment]
I am not sure to understand.... As follow-up of my comment https://github.com/jupyterlab/jupyterlab/pull/12470#issuecomment-1112942395, I have re-enabled the code block with the tracker, and now the settings editor opens correctly. So I guess my JupyterLab was in a "state" (workspace..., tracker...) that was not good, and for some reasons, it made some changes that got it back to a corect state.

**Comment by fcollonval** [comment]
I think we should use a pattern similar to the one advertise in the [extension tutorial](https://jupyterlab.readthedocs.io/en/stable/extension/extension_tutorial.html#restore-panel-state-when-the-browser-refreshes):

```ts
function activate(app: JupyterFrontEnd, palette: ICommandPalette, restorer: ILayoutRestorer) {
  console.log('JupyterLab extension jupyterlab_apod is activated!');

  // Declare a widget variable
  let widget: MainAreaWidget<APODWidget>;

  // Add an application command
  const command: string = 'apod:open';
  app.commands.addCommand(command, {
    label: 'Random Astronomy Picture',
    execute: () => {
      if (!widget || widget.isDisposed) {
        // Create a new widget if one does not exist
        // or if the previous one was disposed after closing the panel
        const content = new APODWidget();
        widget = new MainAreaWidget({content});
        widget.id = 'apod-jupyterlab';
        widget.title.label = 'Astronomy Picture';
        widget.title.closable = true;
      }
      if (!tracker.has(widget)) {
        // Track the state of the widget for later restoration
        tracker.add(widget);
      }
      if (!widget.isAttached) {
        // Attach the widget to the main work area if it's not there
        app.shell.add(widget, 'main');
      }
      widget.content.update();

      // Activate the widget
      app.shell.activateById(widget.id);
    }
  });

  // Add the command to the palette.
  palette.addItem({ command, category: 'Tutorial' });

  // Track and restore the widget state
  let tracker = new WidgetTracker<MainAreaWidget<APODWidget>>({
    namespace: 'apod'
  });
  restorer.restore(tracker, {
    command,
    name: () => 'apod'
  });
}
```

**Comment by echarles** [comment]
Looking around in the various extensions, I can see various patterns for the tracker and setting handling. I am not calling on forging an API for that, but good practices and reviews to ensure they are applied would help.

**Comment by charizard-knows** [comment]
Having the same issue on Win 10 installation. Only difference is that in my case, it never launches

IPython          : 8.4.0
ipykernel        : 6.13.1
ipywidgets       : not installed
jupyter_client   : 7.3.3
jupyter_core     : 4.10.0
jupyter_server   : 1.17.1
jupyterlab       : 3.4.4
nbclient         : 0.6.4
nbconvert        : 6.5.0
nbformat         : 5.4.0
notebook         : 6.4.12
qtconsole        : not installed
traitlets        : 5.2.2

**Comment by fcollonval** [comment]
@charizard-knows would you mind providing any error you are seeing in the browser;  see https://webmasters.stackexchange.com/a/77337 for how to access the JavaScript console

**Comment by charizard-knows** [comment]
@fcollonval Sorry for being so late. This type of delay was unacceptable and I apologize for it.

I am attaching the log of console. I refreshed the browser with the settings tab open (I had another tab open on which I was working)

The error that logs when I actually press
[localhost-1663430751598.log](https://github.com/jupyterlab/jupyterlab/files/9592109/localhost-1663430751598.log)
 the "Json Settings Editor"  begins on line 35

Some background: Running Windows 10 and Opera browser

**Comment by fcollonval** [comment]
Thanks @charizard-knows

We don't official support Opera. Would you mind opening JupyterLab with another browser to see if the bug is appearing or not?

Something else you may try is to reset the workspace. For that when JupyterLab opens in the browser add `?reset` to the URL and hit `Enter`. If that solves the bug, it means it was related to some broken workspace.

**Comment by charizard-knows** [comment]
@fcollonval I attempted the same thing on Firefox but with no difference. Json editor still won't open. Even tried the ?reset workaround but that still doesn't help
Attaching the log
[console-export-2022-9-24_9-26-3.txt](https://github.com/jupyterlab/jupyterlab/files/9637906/console-export-2022-9-24_9-26-3.txt)

**Comment by krassowski** [comment]
@charizard-knows you are hitting an issue unique to JupyterLab 3.4.4 (https://github.com/jupyterlab/jupyterlab/issues/12891) and fixed in 3.4.5 by https://github.com/jupyterlab/jupyterlab/pull/12892. Just upgrade JupyterLab to a newer version and it should go away.

**Comment by charizard-knows** [comment]
@krassowski thanks a lot, upgrading fixed the issue.
